Pass DESKTOP_STARTUP_ID via the socket to focus window

dkondor opened the pull request:
Fixes #3531 at least on KDE Plasma (tested when opening files from Nautilus). On GNOME / Ubuntu there is not change, but things also do not work for Gedit, so that might be an issue on their side.

This works by looking for the `DESKTOP_STARTUP_ID` environment variable when Geany is started. Normally, this is consumed by GTK automatically to use the proper protocols ([Startup notification](https://specifications.freedesktop.org/startup-notification-spec/latest/
) or [xdg-activation](https://wayland.app/protocols/xdg-activation-v1
) on X11 and Wayland respectively). However, when activating an already running Geany instance, this needs to be passed onto it and supplied to GTK there.

This is done by adding a new message to the socket communication for this purpose. This is always parsed, but only has an effect if Geany is running on X11 or Wayland.

Tested (on Ubuntu 24.04):
 - KDE Plasma (kwin 5.27.11): Geany can focus itself with this patch (opening files from Nautilus)
 - GNOME (mutter 46.2): works if Geany is the default for a file type, no change if using the "Open with..." dialog (likely an issue with Nautilus as behavior is the same with other apps, e.g. Gedit)
 - Wayfire: always works, does not care about having a valid activation token
 
Not tested on X11, could potentially improve behavior there as well on WMs that are especially strict with focus stealing prevention.

techee commented:
I can reproduce the problem on Raspberry Pi using Wayland and the default desktop environment and this patch really fixes this issue. 👍 

(On the other hand, with my normal Xfce + X11 Geany gains focus even without this patch.)

dkondor commented:
Good to hear that it works for you.

However, re-reviewing this, I noticed a potential issue: according to the spec, the result of getenv() should not be used after unsetting the same environment variable (even though it seems to work in practice). I've added a workaround that makes a copy instead. Note that with this change, `desktop_startup_id` is not freed if Geany exits from within `parse_command_line_options()` -- if this is an issue, I can think of refactoring further to ensure this.

techee reviewed (changes requested):
I just added some mostly formal review comments but looks good in general.

> Note that with this change, desktop_startup_id is not freed if Geany exits from within parse_command_line_options() -- if this is an issue, I can think of refactoring further to ensure this.

Probably not worth it since Geany terminates anyway and this isn't the normal Geany run which one could e.g. debug with valgrind and see the leak.

techee commented:
Probably fixes also https://github.com/geany/geany/issues/1698

techee commented:
@dkondor Shouldn't also the environment variable `XDG_ACTIVATION_TOKEN` be supported?

https://discourse.gnome.org/t/understanding-gnome-shell-s-focus-stealing-prevention/23434
https://bugzilla.mozilla.org/show_bug.cgi?id=1767546

dkondor commented:
> @dkondor Shouldn't also the environment variable `XDG_ACTIVATION_TOKEN` be supported?
> 
> https://discourse.gnome.org/t/understanding-gnome-shell-s-focus-stealing-prevention/23434 https://bugzilla.mozilla.org/show_bug.cgi?id=1767546

Yes, likely, actually `XDG_ACTIVATION_TOKEN` is the Wayland standard, but it is recommended to use `DESKTOP_STARTUP_ID` as well for interoperability:
https://gitlab.freedesktop.org/wayland/wayland-protocols/-/blob/main/staging/xdg-activation/x11-interoperation.rst?ref_type=heads

Gio seems to set both when using a `GdkAppLaunchContext`, see [here](https://gitlab.gnome.org/GNOME/glib/-/blob/main/gio/gdesktopappinfo.c?ref_type=heads#L2964), but other libraries might use only one of them.

I'll add `XDG_ACTIVATION_TOKEN` as well

dkondor commented on src/libmain.c:
```diff
@@ -1055,6 +1055,14 @@ gint main_lib(gint argc, gchar **argv)
 #ifdef ENABLE_NLS
 	main_locale_init(utils_resource_dir(RESOURCE_DIR_LOCALE), GETTEXT_PACKAGE);
+#endif
+	/* Magic ID used on X11 and Wayland for bringing our existing window on top;
```
Sure, done

dkondor commented on src/libmain.c:
```diff
@@ -1055,6 +1055,14 @@ gint main_lib(gint argc, gchar **argv)
+	 * unsetenv() (which is done by GTK). */
+#ifdef HAVE_SOCKET
+	gchar *desktop_startup_id = g_strdup(getenv("DESKTOP_STARTUP_ID"));
 #endif
```
Done

dkondor commented on src/libmain.c:
```diff
@@ -1085,7 +1093,7 @@ gint main_lib(gint argc, gchar **argv)
 		socket_info.lock_socket = -1;
 		socket_info.lock_socket_tag = 0;
-		socket_info.lock_socket = socket_init(argc, argv, socket_port);
+		socket_info.lock_socket = socket_init(argc, argv, socket_port, desktop_startup_id);
```
Unfortunately, this would leak if `socket_info.ignore_socket == FALSE`

dkondor commented on src/socket.c:
```diff
@@ -124,13 +126,21 @@ static gint socket_fd_close			(gint sock);
 
+	if (desktop_startup_id != NULL)
+	{
+		geany_debug("sending DESKTOP_STARTUP_ID before file open command");
```
Makes sense, moved to later

dkondor commented on src/socket.c:
```diff
@@ -719,6 +729,22 @@ gboolean socket_lock_input_cb(GIOChannel *source, GIOCondition condition, gpoint
+			while (socket_fd_gets(sock, buf, sizeof(buf)) != -1 && *buf != '.')
+			{
+				g_strstrip(buf); /* remove \n char */
+#if defined(GDK_WINDOWING_WAYLAND) && GTK_CHECK_VERSION(3, 22, 0)
```
Done

dkondor commented on src/socket.c:
```diff
@@ -719,6 +729,22 @@ gboolean socket_lock_input_cb(GIOChannel *source, GIOCondition condition, gpoint
+		else if (strncmp(buf, "desktop_startup_id", 18) == 0)
+		{
+			GdkDisplay *display = gdk_display_get_default();
+			while (socket_fd_gets(sock, buf, sizeof(buf)) != -1 && *buf != '.')
```
The main reason for the while loop is to make sure we get to the end of the message (we should not have a line break in the token, but better be safe); I've refactored this to really only use the first line

techee commented on src/socket.c:
```diff
@@ -719,6 +728,30 @@ gboolean socket_lock_input_cb(GIOChannel *source, GIOCondition condition, gpoint
+					gdk_x11_display_set_startup_notification_id(display, buf);
+#endif
+#endif // GDK_WINDOWING_WAYLAND || GDK_WINDOWING_X11
+				while (socket_fd_gets(sock, buf, sizeof(buf)) != -1 && *buf != '.')
```
Possibly good indeed if tokens contain newlines (probably not).

techee commented on src/socket.c:
```diff
@@ -719,6 +729,22 @@ gboolean socket_lock_input_cb(GIOChannel *source, GIOCondition condition, gpoint
 		}
+		else if (strncmp(buf, "desktop_startup_id", 18) == 0)
+		{
+			GdkDisplay *display = gdk_display_get_default();
```
It probably doesn't cause any harm if we keep it.

techee reviewed (changes requested):
Looks good apart from the minor code style comment.

Just curious, didn't the use of `XDG_ACTIVATION_TOKEN` fix the focus issue under gnome by any chance?

techee commented on src/socket.c:
```diff
@@ -719,6 +728,30 @@ gboolean socket_lock_input_cb(GIOChannel *source, GIOCondition condition, gpoint
+			{
+#if defined(GDK_WINDOWING_WAYLAND) || defined(GDK_WINDOWING_X11)
+				g_strstrip(buf); /* remove \n char */
+				geany_debug("got desktop_startup_id: %s", buf);
```
Maybe also, could the string be modified to something similar to https://github.com/geany/geany/blob/22a8a1f55e30db8566f752211bcd895c707ea24c/src/socket.c#L678 so it's a little more descriptive? E.g. something like

```
geany_debug("Received IPC command from remote instance: desktop_startup_id (%s)", buf)
```

techee commented:
I also tested it on Ubuntu 24.04 and this fixes the issue there as well 🎉

dkondor commented:
I addressed the remaining comments as well.

I also re-tested and now it works for me as well on Ubuntu 24.04 when Geany is the default for a file. What I found is that it sometimes does not work if selecting it from the "Open with..." dialog, but I think this is likely a bug in Nautilus (as it is the same behavior with other apps as well such as Gedit). I'll investigate and report it to them separately.

So yes, this should also fix #1698 as well which seems to be the same problem.

Should I squash the commits in this branch? (I think logically they belong together)

techee commented:
> Should I squash the commits in this branch? (I think logically they belong together)

Yes please, this PR is done from my point of view. Let's see if others have any remarks.

techee commented:
IMO this patch works well and is pretty straightforward so unless there are any objections, I'll merge it in about a week.
